### PR TITLE
fix existing MantleBackup's reconciler for data replication

### PR DESCRIPTION
**Pull request by ushitora-anqou**

### CODE REVIEW

File: internal/controller/mantlebackup_controller.go [hunk 1]
```diff
@@ -400,10 +406,10 @@ func (r *MantleBackupReconciler) replicate(
...
 	}
 	for _, backup1 := range backupList.Items {
-		cond := meta.FindStatusCondition(backup1.Status.Conditions, mantlev1.BackupConditionSyncedToRemote)
-		if (backup1.Status.SnapID == nil || *backup1.Status.SnapID < *backup.Status.SnapID) &&
-			backup1.ObjectMeta.DeletionTimestamp.IsZero() &&
-			(cond == nil || cond.Status != metav1.ConditionTrue) {
+		if backup1.Status.SnapID == nil ||
+			*backup1.Status.SnapID < *backup.Status.SnapID &&
+				backup1.ObjectMeta.DeletionTimestamp.IsZero() &&
+				!meta.IsStatusConditionTrue(backup1.Status.Conditions, mantlev1.BackupConditionSyncedToRemote) {
```
Comment (satoru-takeuchi) [inline internal/controller/mantlebackup_controller.go:412]: The condition is OK. However, this situation is not an error like the failure of access to API servers but waiting for fullfilling some conditions. So, it's better to return from `Reconcile()`  with `ctrl.Result{Requeue: true}, nil`. Probably the return value of `replicate()` should also be changed to distinguish whether the request will be requeued due to error or not.

I should be noticed this point in the previous PR which is to implementing `replicate()`. My fault.

**Review by satoru-takeuchi — CHANGES_REQUESTED** [review]

File: internal/controller/mantlebackup_controller.go [hunk 2]
```diff
@@ -366,7 +366,13 @@ func (r *MantleBackupReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	}
 
 	if r.role == RolePrimary {
-		if err := r.replicate(ctx, logger, &backup, target.pvc); err != nil {
+		// Unmarshal the PVC manifest stored in the status of the MantleBackup resource.
+		var pvc corev1.PersistentVolumeClaim
+		if err := json.Unmarshal([]byte(backup.Status.PVCManifest), &pvc); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to unmarshal the PVC stored in the status of the MantleBackup resource: %w", err)
+		}
```
Comment (toshipp) [inline internal/controller/mantlebackup_controller.go:373]: Writing logic in Reconcile method makes it hard to read and maintain, IMO.
I'd like to embed it into replicate method as a precondition check or split it as a function.

**Comment by ushitora-anqou** [comment]
@satoru-takeuchi @toshipp I fixed my code according to your reviews. Could you review this PR again? The changes are as follows: https://github.com/cybozu-go/mantle/compare/6d291c37d39095bca8d6a54db12329b5bd67aa63..a5039dd0cda8ae26cea5ce2343926c5881b49b0c

File: internal/controller/mantlebackup_controller.go [hunk 3]
```diff
@@ -366,7 +366,9 @@ func (r *MantleBackupReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	}
 
 	if r.role == RolePrimary {
-		if err := r.replicate(ctx, logger, &backup, target.pvc); err != nil {
+		if err := r.replicate(ctx, logger, &backup); err == errSkipProcessing {
```
Comment (toshipp) [inline internal/controller/mantlebackup_controller.go:369]: My intent of errSkipProcessing is to finish reconciling the MB without requeue. If you want requeue, why don't you return ctrl.Result from replicate method?

Comment (ushitora-anqou) [reply internal/controller/mantlebackup_controller.go]: I fixed my code. Could you take another look at it to make sure it's what you had in mind?

**Review by toshipp — COMMENTED** [review]
lgtm

**Review by satoru-takeuchi — APPROVED** [review]

**Comment by ushitora-anqou** [comment]
Thank you!

**Pull request merged**